### DISCUSSION TITLE
Support of server components in Next.js

### DISCUSSION THREAD

**Discussion by shemayas** [ben-rogerson/twin.examples#122] [Q&A]
Is there support for server components in Next.js?
I'm trying to use it with [this example](https://github.com/ben-rogerson/twin.examples/tree/master/next-emotion-typescript) using the app directory instead of pages, but I'm getting an error:
![image](https://github.com/ben-rogerson/twin.examples/assets/11064714/8aaa4769-4254-4583-9ff3-a00f5d6c903c)

**Comment by ullaskunder3** [comment]
this is causing because you are trying to use context hooks in Next.js which is by default uses SSR,

if you still want to use `createContext` then you might have to take different approach
- 1st option you can make it a `CSR` client side render by using `"use client"` at the top
- 2nd option you can create a separate file where you should implement the your `context` code with "use client" at the top 

## example
```jsx
"use client";

import { createContext, useContext } from 'react';
import { useSession } from '../lib/useSession';

export const AppUserContext = createContext({
  session: null,
  profile: null
});

export default function AppUserProvider({ children }) {
  const appUserInfo = useSession();
  return (
    <AppUserContext.Provider value={appUserInfo}>
      <>{children}</>
    </AppUserContext.Provider>
  );
}

export const useAppUserGlobalContext = () => {
  const context = useContext(AppUserContext);

  if (context === undefined) {
    throw new Error('data must be used inside AppUserProvider');
  }

  return context;
};

```
 - export it then import `AppUserProvider` in layout.tsx
 ```
         <AppUserProvider>
          <div className="select-none dark:bg-slate-800/80">
            <Header />
              {children}
          </div>
        </AppUserProvider>
 ```
  - and where you want to use the context import `useAppUserGlobalContext` *have to use `use client` in that page.tsx*
  ```
  import { useSupabaseGlobalContext } from "../../hooks/supabaseProvider";

  ```
